### PR TITLE
Multicall

### PR DESCRIPTION
**Pull request by dmcrodrigues**
### Dynamic arrays with dynamic types

Adds support to handle correctly dynamic arrays with a dynamic type like `bytes[]` which require a specific decoding.

```
* offset (64 nibbles)
* array length (64 nibbles)
* offsets for each element of the array (64 nibbles)
* size of 1st element of the array (64 nibbles)
* 1st element of the array (dynamic)
* size of 2nd element of the array (64 nibbles)
* 2nd element of the array (dynamic)
...
* size of N-th element of the array (64 nibbles)
* N-th element of the array (dynamic)
```

### Multicall Contract

Multicall contract allows batching multiple function calls into a single one. The contract allows fetching the name, symbol, decimals and balance for a particular token (or tokens) in a single function call instead of having 4 separate function calls for each component (multiplied by the number of tokens).

```swift
var aggregator = Multicall.Aggregator()

// Handle local result in-place
try aggregator.append(
    function: ERC20Functions.name(contract: testContractAddress),
    response: ERC20Responses.nameResponse.self
) { result in
    ...
}

// Handle results in the global handler
try aggregator.append(ERC20Functions.symbol(contract: testContractAddress))


// Execute
multicall.aggregate(calls: aggregator.calls) { result in
    // response.outputs[0] holds the result from the 1st call, `ERC20Functions.name`
    let symbol = try ERC20Responses.symbolResponse(data: try response.outputs[1].get())?.value
    ...
}
```

The contract is based on [Make's Multicall](https://github.com/makerdao/multicall) but was modified to allow failures in individual calls that are flagged through a specific error to consumers.

#### Deployment

| Network | Address |
| ------------- | ------------- |
| Ropsten  | [0x604D19Ba889A223693B0E78bC1269760B291b9Df](https://ropsten.etherscan.io/address/0x604D19Ba889A223693B0E78bC1269760B291b9Df) |
| Mainnet  | [0xF34D2Cb31175a51B23fb6e08cA06d7208FaD379F](https://etherscan.io/address/0xF34D2Cb31175a51B23fb6e08cA06d7208FaD379F) |

### CODE REVIEW

File: web3sTests/Multicall/MulticallTests.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        try aggregator.append(ERC20Functions.symbol(contract: testContractAddress))
+
+        let expect = expectation(description: "Get token name and symbol")
+        multicall.aggregate(calls: aggregator.calls) { result in
+            do {
+                switch result {
+                case .failure(let error):
+                    XCTFail("Multicall failed with error: \(error)")
+                case .success(let response):
+                    let name = try response.outputs[0].value
```
Comment (dmcrodrigues) [inline web3sTests/Multicall/MulticallTests.swift:36]: I'm evaluating how to improve the way we pass the output to consumers to allow a better handling as right now they need to correlate the order of each call with the expected output and do the respective parsing of the response.

**Review by DarthMike — APPROVED** [review]

File: web3sTests/Multicall/MulticallTests.swift (hunk 1 above)
Comment (dmcrodrigues) [reply web3sTests/Multicall/MulticallTests.swift]: Improved in https://github.com/argentlabs/web3.swift/pull/116/commits/47cede40dd314180c09b490961d9355f64e1427e.

**Review by DarthMike — APPROVED** [review]

**Review by DarthMike — APPROVED** [review]
Not sure why tests not passing anymore, we'll need to review them before merging, but seems it's ropsten issue, not code change regression

**Comment by DarthMike** [comment]
Tests passing locally, so must be temporary issue with Ropsten

File: web3sTests/ENS/ENSTests.swift [hunk 2]
```diff
@@ -126,5 +126,38 @@ class ENSTests: XCTestCase {
         
         waitForExpectations(timeout: 20)
     }
-    
+
+    func testGivenRopstenRegistry_TODO() {
```
Comment (DarthMike) [inline web3sTests/ENS/ENSTests.swift:130]: `testGiven3Addresses_whenOneDoesNotResolve_ReturnsCorrectValues` ?

**Review by DarthMike — APPROVED** [review]

File: web3sTests/ENS/ENSTests.swift [hunk 3]
```diff
@@ -126,5 +126,72 @@ class ENSTests: XCTestCase {
...
+            results,
+            [
+                .resolved("julien.argent.test"),
+                .couldNotBeResolved(.ensUnknown),
+                .resolved("davidtests.argent.xyz")
+            ]
+        )
+    }
+
+    func testGivenRopstenRegistry_TODO2() {
```
Comment (DarthMike) [inline web3sTests/ENS/ENSTests.swift:164]: `testGiven3EBS_whenOneDoesNotResolve_ReturnsCorrectValues` ?

File: web3sTests/ENS/ENSTests.swift (hunk 2 above)
Comment (dmcrodrigues) [reply web3sTests/ENS/ENSTests.swift]: Fixed in https://github.com/argentlabs/web3.swift/pull/116/commits/8495db05bddf4235bce8c3e4a055d333937b1a87.

File: web3sTests/ENS/ENSTests.swift (hunk 3 above)
Comment (dmcrodrigues) [reply web3sTests/ENS/ENSTests.swift]: Fixed in https://github.com/argentlabs/web3.swift/pull/116/commits/8495db05bddf4235bce8c3e4a055d333937b1a87.

**Review by th3m477 — APPROVED** [review]

**Pull request merged**